transport/mixer/src/channel.rs

```diff
@@ -0,0 +1,487 @@
+        let is_active = self.channel.receiver_active.load(Ordering::Relaxed);
+        if is_active {
+            Poll::Ready(Ok(()))
+        } else {
+            Poll::Ready(Err(SenderError::Closed))
+        }
+    }
+
+    #[tracing::instrument(level = "trace", skip(self, item))]
+    fn start_send(self: std::pin::Pin<&mut Self>, item: T) -> Result<(), Self::Error> {
+        let is_active = self.channel.receiver_active.load(Ordering::Relaxed);
+
+        if is_active {
+            let mut channel = self.channel.channel.lock().map_err(|_| SenderError::Lock)?;
+
+            let random_delay = channel.cfg.random_delay();
+
+            trace!(delay_in_ms = random_delay.as_millis(), "generated mixer delay",);
+
+            let delayed_data: DelayedData<T> = (std::time::Instant::now() + random_delay, item).into();
+            channel.buffer.push(Reverse(delayed_data));
+
+            if let Some(waker) = channel.waker.as_ref() {
+                waker.wake_by_ref();
+            }
+
+            #[cfg(all(feature = "prometheus", not(test)))]
+            {
+                METRIC_QUEUE_SIZE.increment(1.0f64);
+
+                let weight = 1.0f64 / channel.cfg.metric_delay_window as f64;
+                METRIC_MIXER_AVERAGE_DELAY.set(
+                    (weight * random_delay.as_millis() as f64) + ((1.0f64 - weight) * METRIC_MIXER_AVERAGE_DELAY.get()),
+                );
+            }
+
+            Ok(())
+        } else {
+            Err(SenderError::Closed)
+        }
+    }
+
+    fn poll_flush(self: std::pin::Pin<&mut Self>, _cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: std::pin::Pin<&mut Self>, _cx: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // The channel can only be closed by the receiver. The sender can be dropped at any point.
+        Poll::Ready(Ok(()))
+    }
+}
+
+/// Error returned by the [`Receiver`].
+#[derive(Debug, thiserror::Error)]
+pub enum ReceiverError {
+    /// The channel is closed due to receiver being dropped.
+    #[error("Channel is closed")]
+    Closed,
+
+    /// The mutex lock over the channel failed.
+    #[error("Channel lock failed")]
+    Lock,
+}
+
+/// Receiver object interacting with the mixer channel.
+///
+/// The receiver receives already mixed elements without any knowledge of
+/// the original order.
+pub struct Receiver<T> {
+    channel: TrackedChannel<T>,
+}
+
+impl<T> Stream for Receiver<T> {
+    type Item = T;
+
+    #[tracing::instrument(level = "trace", skip(self, cx))]
+    fn poll_next(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Option<Self::Item>> {
+        let now = std::time::Instant::now();
+        if self.channel.sender_count.load(Ordering::Relaxed) > 0 {
+            let mut channel = self.channel.channel.lock().unwrap();
```

Comment:
This is what `let-else` clause is for. But it would be better once your stream yields `Result`.
```suggestion
            let Ok(mut channel) = self.channel.channel.lock() else {
                              tracing::error!("mutex is poisoned, terminating stream");
                              return Poll::Ready(None);
            }
```